### DISCUSSION TITLE
Virtual Machine Standard Library Features

### DISCUSSION THREAD

**Discussion by hrszpuk** [hrszpuk/rvm#33] [General]
### What libraries should be in the standard library?
Non-ticked entries are just proposals. Ticked entries likely to be implemented. 
- [x] `io` - I/O for streams
- [x] `threads` - Multithreading API
- [x] `mem` - Dynamic memory management
- [x] `math` - Mathematical functions
- [x] `random` - Generating pseudo-random numbers
- [x] `os` - Miscellaneous operating system interfaces
- [x] `sys` - System-specific parameters and functions
- [ ] `gc` - Garbage collection interface
- [ ] `vm` - Virtual machine data
- [ ] `sockets` - Low-level networking interface
- [ ] `ssl` - TLS/SSL wrapper for sockets
- [x] `strings` - Strings utility library
- [x] `ldex` - Dynamic Library Loading

### How will the standard library work?
The virtual machine contains all standard library code within its configuration library.
When building your code with `rvm build`, the translator will automatically link with standard libraries if they are imported into the code.

```asm
.meta
.extern <io.rvmt> io ; This is a standard library package. They are included using <> instead of "".
; NOTE(hrs): both <io.rvmt> and <io> should be accepted.

.data
.const message "Hello, World"

.code
push message
call [io]print
halt
```
Most standard library packages will be calling functions from C.

**Comment by Khhs167** [comment]
Could `reflection` be a possibility? Maybe it'd be part of `vm`.

Thinking that it's useful to be able to read data from libraries within RVM

**Reply by hrszpuk** [reply]
I was thinking `vm` would be for accessing data about the VM. So, if you, for what ever reason, wanted to see what was in the current call frame, you'd use `call [vm]GetCurrentCallFrame` or smth.
